File: generator/generator.rb
#!/usr/bin/env ruby
#
# ruby generator.rb <num of strings>
#
require 'securerandom'
require 'ffi'

module GoUuid
  extend FFI::Library
    ffi_lib './gouuid/goUuid.so'
      attach_function :GoUuid, [], :string
end

amount = ARGV[0]
amount ||= 100
p "Generating " + amount.to_s + " lines..."

start = Time.now

#number of different strings
num_diff = rand(3..20)
#array of index of diff strings
arr_num_diff = []
num_diff.times do |i|
  arr_num_diff << rand(1..amount.to_i)
end
p arr_num_diff

out = []
amount.to_i.times do |i|
  #out << SecureRandom.uuid
  out << GoUuid.GoUuid()
end

p out.size
f1 = File.open("../example/first.txt", "w")
f2 = File.open("../example/second.txt", "w")

out.each.with_index do |el, i|
  f1 << el + "\n"
  next if arr_num_diff.include? i
  f2 << el + "\n"
end

f1.close
f2.close

p Time.now - start

File: generator/generator.go
package main

import (
	"fmt"
	"github.com/satori/go.uuid"
	"math/rand"
	"os"
	"runtime"
	"sort"
	"strconv"
	"time"
)

func writeFiles(toDel []int, arr []string) {
	f1, err := os.Create("../example/first.txt")
	defer f1.Close()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	f2, err := os.Create("../example/second.txt")
	defer f2.Close()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	for i, el := range arr {
		_, err := f1.WriteString(el + "\n")
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		if excludeString(i, toDel) {
			continue
		}
		_, err = f2.WriteString(el + "\n")
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
	}
}

func genRandomNumbers(val int) []int {
	rand.Seed(time.Now().UnixNano())
	arr := []int{}
	// exclude random number of random numbers
	el := rand.Intn(20)
	for i := 0; i < el; i++ {
		arr = append(arr, rand.Intn(val))
	}
	return arr
}

func excludeString(el int, arr []int) bool { // arr must be sorted!
	for i := 0; i < len(arr); i++ {
		if el < arr[i] {
			break
		}
		if el == arr[i] {
			return true
		}
	}
	return false
}

func main() {
	runtime.GOMAXPROCS(4)
	if len(os.Args) <= 1 {
		fmt.Println("Usage: generator <num of uuids>")
		os.Exit(1)
	}
	start := time.Now()
	// Creating UUID Version 4
	val, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	toDel := genRandomNumbers(val)
	sort.Ints(toDel)
	fmt.Printf("Exclude %d elements\n", len(toDel))
	arr := []string{}
	for i := 0; i < val; i++ {
		arr = append(arr, fmt.Sprintf("%s", uuid.NewV4()))
	}
	writeFiles(toDel, arr)
	fmt.Println(time.Since(start))
}
